renderdoc/driver/ihv/amd/amd_counters.cpp

```diff
@@ -391,26 +445,89 @@ uint32_t AMDCounters::GetPassCount()
   return (uint32_t)numRequiredPasses;
 }
 
-uint32_t AMDCounters::BeginSession()
+uint32_t AMDCounters::CreateSession()
 {
-  gpa_uint32 sessionID = 0;
+  uint32_t sessionID = m_gpaSessionCounter;
+  GPA_SessionId gpaSessionId = nullptr;
+  GPA_Status status = m_pGPUPerfAPI->GPA_CreateSession(
+      m_gpaContextId, GPA_SESSION_SAMPLE_TYPE_DISCRETE_COUNTER, &gpaSessionId);
 
-  GPA_Status status = m_pGPUPerfAPI->GPA_BeginSession(&sessionID);
   if(AMD_FAILED(status))
   {
-    GPA_ERROR("Begin session.", status);
+    GPA_ERROR("Create session.", status);
+  }
+  else
+  {
+    InitializeCmdInfo();
+    m_gpaSessionInfo.push_back(gpaSessionId);
+    ++m_gpaSessionCounter;
   }
 
-  return (uint32_t)sessionID;
+  return sessionID;
+}
+
+void AMDCounters::BeginSession(uint32_t sessionId)
+{
+  GPA_Status status = m_pGPUPerfAPI->GPA_BeginSession(m_gpaSessionInfo.at(sessionId));
+  if(AMD_FAILED(status))
+  {
+    GPA_ERROR("Begin session.", status);
+  }
+  else
+  {
+    m_passCounter = -1;
+  }
 }
 
-void AMDCounters::EndSesssion()
+void AMDCounters::EndSesssion(uint32_t sessionId)
 {
-  GPA_Status status = m_pGPUPerfAPI->GPA_EndSession();
+  GPA_Status status = m_pGPUPerfAPI->GPA_EndSession(m_gpaSessionInfo.at(sessionId));
   if(AMD_FAILED(status))
   {
     GPA_ERROR("End session.", status);
   }
+
+  m_passCounter = 0u;
+}
+
+void AMDCounters::InitializeCmdInfo()
+{
+  switch(m_apiType)
+  {
+    case ApiType::Dx11:
+    case ApiType::Ogl: break;
+    case ApiType::Vk:
+    case ApiType::Dx12:
+      if(nullptr == m_gpaCmdListInfo.m_pCommandListMap)
+      {
+        m_gpaCmdListInfo.m_pCommandListMap = new std::map<void *, GPA_CommandListId>();
+      }
+      break;
+  }
+}
+
+void AMDCounters::DeInitializeCmdInfo()
+{
+  switch(m_apiType)
+  {
+    case ApiType::Dx11:
+    case ApiType::Ogl: break;
+    case ApiType::Vk:
+    case ApiType::Dx12:
+      delete m_gpaCmdListInfo.m_pCommandListMap;
+      m_gpaCmdListInfo.m_pCommandListMap = nullptr;
```

Comment:
`NULL` rather than `nullptr` here please. There's also a `SAFE_DELETE` macro which does this all together.

Reply:
Updated